### DISCUSSION TITLE
Unique indexes do not seem to work in combination with document inheritance

### DISCUSSION THREAD

**Discussion by sensoryoverload** [JasperFx/marten#1421] [Q&A] [answered]
When using this combination the unique index is not created in the database. However, creating an instance of the DocumentStore does not complain about this inability. I might be missing something though as I am just started using Marten.

```using Marten;
using Marten.Schema.Identity;
using System;

namespace MartenComputedIndex
{
    class JobPosition
    {
        public Guid Id { get; set; }
        public int ExternalKey { get; set; }
    }

    class ExternalJobPosition : JobPosition
    {
        public ExternalJobPosition(string title, int externalKey)
        {
            ExternalKey = externalKey;
        }
    }

    public class Program
    {
        static void Main(string[] args)
        {
            using (var store = StartDocumentStore())
            {
                // Clear all data from postgresql tables
                store.Advanced.Clean.CompletelyRemoveAll();

                using (var session = store.OpenSession())
                {
                    // Insert first job position
                    session.Store(new ExternalJobPosition(title: "Job 1", externalKey: 1));
                    session.SaveChanges();

                    // Insert second job position with another external key
                    session.Store(new ExternalJobPosition(title: "Job 2", externalKey: 2));
                    session.SaveChanges();

                    // Save changes should throw because externalKey: 1 is already used in another job position
                    session.Store(new ExternalJobPosition(title: "Job 1", externalKey: 1));
                    session.SaveChanges();
                }
            }
        }

        static DocumentStore StartDocumentStore()
        {
            return DocumentStore.For(_ =>
            {
                _.Connection("connectionstring");
                _.AutoCreateSchemaObjects = AutoCreate.All;
                _.DefaultIdStrategy = (mapping, storeOptions) => new CombGuidIdGeneration();
                _.UseDefaultSerialization(nonPublicMembersStorage: NonPublicMembersStorage.NonPublicSetters);

                // Define inheritance structure (Removing this line will result in a unique index)
                _.Schema.For<JobPosition>().AddSubClass<ExternalJobPosition>();

                // Using a computed index does not seem to work
                _.Schema.For<ExternalJobPosition>().UniqueIndex(x => x.ExternalKey);

                // Using a duplicated index does not seem to work either
                //_.Schema.For<ExternalJobPosition>().UniqueIndex(UniqueIndexType.DuplicatedField, x => x.ExternalKey);
            });
        }
    }
}
```

**Comment by sensoryoverload** [comment]
It does seem to work if I put the UniqueIndex on the class where the property is defined. But only when that class is the base of the inheritance structure, using it on a subclass still wont work. If I move the property to be indexed to a subclass it wont work

```
//_.Schema.For<JobPosition>().UniqueIndex(x => x.ExternalKey);
_.Schema.For<ExternalJobPosition>().UniqueIndex(x => x.ExternalKey);
```

**Comment by mysticmind** [comment, marked as answer]
@sensoryoverload Sorry about the delayed reply. Few things to note when you use document hierarchies/inheritance:
- `_.Schema.For<JobPosition>().AddSubClass<ExternalJobPosition>();`, this line tells Marten that these classes have document hierarchies 
- All the documents pertaining to the hierarchies gets stored in one table and Marten provides you the ability to query on base class or sub classes. When defining any indexes including unique indexes, it can be applied only on base classes (you have already observed as the behavior).
- There are few limitations when would want to use the document hierarchies and unique indexes on sub classes.

If you **did not** register document hierarchies i.e. `_.Schema.For<JobPosition>().AddSubClass<ExternalJobPosition>();` then Marten will treat `JobPosition` and `ExternalJobPosition` to be stored as documents in separate tables. In this case, when you define unique indexes, it gets added to different tables.

I think we have to update the docs to add some clarity around the limitations of using unique indexes on sub classes in the document hierarchies.

**Comment by sensoryoverload** [comment]
Thanks for the response. In retrospect, I should have known that what I tried would not work.
